lib/kino/explorer.ex

```diff
@@ -0,0 +1,107 @@
+defmodule Kino.Explorer do
+  @moduledoc """
+  A widget for interactively viewing `Explorer.DataFrame`.
+
+  ## Examples
+
+      df = Explorer.Datasets.fossil_fuels()
+      Kino.Explorer.new(df)
+  """
+
+  use GenServer, restart: :temporary
+
+  alias Kino.Utils.Table
+
+  defstruct [:pid]
+
+  @type t :: %__MODULE__{pid: pid()}
+
+  @typedoc false
+  @type state :: %{
+          parent_monitor_ref: reference(),
+          df: Explorer.DataFrame.t(),
+          total_rows: non_neg_integer()
+        }
+
+  @doc """
+  Starts a widget process with a data frame.
+  """
+  @spec new(Explorer.DataFrame.t()) :: t()
+  def new(df) do
+    parent = self()
+    opts = [df: df, parent: parent]
+
+    {:ok, pid} = DynamicSupervisor.start_child(Kino.WidgetSupervisor, {__MODULE__, opts})
+
+    %__MODULE__{pid: pid}
+  end
+
+  @doc false
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts)
+  end
+
+  @impl true
+  def init(opts) do
+    df = Keyword.fetch!(opts, :df)
+    parent = Keyword.fetch!(opts, :parent)
+
+    parent_monitor_ref = Process.monitor(parent)
+
+    total_rows = Explorer.DataFrame.n_rows(df)
+    keys = Explorer.DataFrame.names(df)
+
+    {:ok,
+     %{
+       parent_monitor_ref: parent_monitor_ref,
+       df: df,
+       keys: keys,
+       total_rows: total_rows
+     }}
+  end
+
+  @impl true
+  def handle_info({:connect, pid}, state) do
+    columns = Table.keys_to_columns(state.keys)
+    features = [:pagination, :sorting]
+
+    send(pid, {:connect_reply, %{name: "DataFrame", columns: columns, features: features}})
+
+    {:noreply, state}
+  end
+
+  def handle_info({:get_rows, pid, rows_spec}, state) do
+    records = get_records(state.df, rows_spec)
+    rows = Enum.map(records, &Table.record_to_row(&1, state.keys))
+
+    send(pid, {:rows, %{rows: rows, total_rows: state.total_rows, columns: :initial}})
+
+    {:noreply, state}
+  end
+
+  def handle_info({:DOWN, ref, :process, _object, _reason}, %{parent_monitor_ref: ref} = state) do
+    {:stop, :shutdown, state}
+  end
+
+  defp get_records(df, rows_spec) do
+    df =
+      if order_by = rows_spec[:order_by] do
+        Explorer.DataFrame.arrange(df, [{order_by, rows_spec.order}])
```

Comment:
In particular, this will change to `[{rows_spec.order, order_by}]` soon as far as I understand.

Reply:
That's right! `order_by` is a `List` in either case.

Reply:
I'm actually going to more closely match the `Ecto.Query` behaviour as José suggested, so it won't be a list. PR to follow.

Reply:
IIRC the point was to replace `name: :asc, age: :desc` by `asc: "name", desc: "age"`, so it's still a keyword list? For Ecto we do:

https://github.com/livebook-dev/kino/blob/c38ee8750361cc0a586b67049a5432db0fd1dfe0/lib/kino/ecto.ex#L157-L158

Reply:
That's right. I was saying it could be a keyword list with the value as a list so: `Keyword.t([String.t()])`, e.g. `asc: ["name1", "name2"], desc: ["age"]` but went with the way you described it above. See elixir-nx/explorer#39.

Reply:
Great! Updated here accordingly :)